File: src/1124-Longest-Well-Performing-Interval/1124.cpp
class Solution 
{
public:
    int longestWPI(vector<int>& hours) 
    {
        int res = 0, pre_sum = 0;
        unordered_map<int, int> seen;
        for (int i = 0; i < hours.size(); ++i)
        {
            pre_sum += hours[i] > 8 ? 1 : -1;
            if (!seen.count(pre_sum)) seen[pre_sum] = i;
            if (pre_sum > 0) res = i + 1;
            else if (seen.count(pre_sum - 1)) res = max(res, i - seen[pre_sum - 1]);
        }
        return res;
    }
};

File: src/1124-Longest-Well-Performing-Interval/1124.py
class Solution:
    def longestWPI(self, hours: List[int]) -> int:
        res = pre_sum = 0
        seen = {}
        
        for i, h in enumerate(hours):
            pre_sum += (1 if h > 8 else -1)
            if pre_sum not in seen:
                seen[pre_sum] = i
            if pre_sum > 0:
                res = i + 1
            elif pre_sum - 1 in seen:
                res = max(res, i - seen[pre_sum - 1])
        return res